src/toil/lib/threading.py

```diff
@@ -35,6 +36,52 @@
 
 logger = logging.getLogger(__name__)
 
+def safe_lock(fd: int, block: bool = True, shared: bool = False) -> None:
+    """
+    Get an fcntl lock, while retrying on IO errors.
+
+    Raises OSError with EACCES or EAGAIN when a nonblocking lock is not
+    immediately available.
+    """
+
+    error_backoff = 1
+
+    while True:
+        try:
+            # Wait until we can exclusively lock it.
+            fcntl.lockf(fd, (fcntl.LOCK_SH if shared else fcntl.LOCK_EX) | (fcntl.LOCK_NB if not block else 0))
+            return
+        except OSError as e:
+            if e.errno in (errno.EACCES, errno.EAGAIN):
+                # Nonblocking lock not available.
+                raise
+            elif e.errno == errno.EIO:
+                # Sometimes Ceph produces IO errors when talking to lock files.
+                # Back off and try again.
+                # TODO: Should we eventually give up if the disk really is
+                # broken? If so we should use the retry system.
+                logger.error("IO error talking to lock file. Retrying after %s seconds.", error_backoff)
+                time.sleep(error_backoff)
+                error_backoff = min(60, error_backoff * 2)
+                continue
+            else:
+                raise
+
+def safe_unlock_and_close(fd: int) -> None:
+    """
+    Release an fcntl lock and close the file descriptor, while handling fcntl IO errors.
+    """
+    try:
+        fcntl.lockf(fd, fcntl.LOCK_UN)
+    except OSError as e:
+        if e.errno == errno.EIO:
+            # Sometimes Ceph produces these. We don't need to retry because
+            # we're going to close the FD and after that the file can't remain
+            # locked by us.
+            pass
+        else:
+            raise
+    os.close(fd)
```

Comment:
Suggestion:

```
    try:
        fcntl.lockf(fd, fcntl.LOCK_UN)
    except OSError as e:
        if e.errno != errno.EIO:
            # Sometimes Ceph produces errno.EIO .  We don't need to retry because
            # we're going to close the FD and after that the file can't remain
            # locked by us.
            raise
    os.close(fd)
```

Reply:
This doesn't leave the comment really in the right place relative to the code.